### DISCUSSION TITLE
'proj' object in the Quick Tutorial

### DISCUSSION THREAD

**Discussion by sanghoonleeeins** [GreenleafLab/ArchR#1015] [Questions / Documentation] [answered]
Hi, 

I am reading the quick tutorial, https://www.archrproject.com/articles/Articles/tutorial.html, and following the code. As I am a beginner, I am confused in several lines. 

In the beginnig, I made 'proj' object by 'ArchRProject' function
> proj<-ArchRProject(ArrowFiles=ArrowFiles, outputDirectory="HemeTutorial",
  copyArrows = TRUE #This is recommened so that you maintain an unaltered copy for later usage.
)

Then, I filter out doublects by 'filterDoublets' function.  Then, it gets the output to 'proj' object, which is **the same as ArchRProject.**  This makes me confused when I run the down streams. 
> proj<-filterDoublets(ArchRProj=proj)

Next step is dimensionality reduction. 
> proj<- **addIterativeLSI(ArchRProj=proj**, useMatrix = "TileMatrix", name = "IterativeLSI")
> proj<-addClusters(input=**proj**, reducedDims = "IterativeLSI")

In 'addIterativeLSI', ArchRProj=**proj**  will be from 'proj' by filterDoublets if I want to use 'proj' after filtering doublets, am I correct?  
In 'addClusters', input=**proj** also will be from 'proj' by filterDoublets, not by "addIterativeLSI" which is in the just above line.  Could you confirm whether I understood correctly?    First time, I was confused and put the 'input=proj' that I got by 'addIterativeLSI' in the just above line.

I was confused because many functions, such as ArchRProject, filterDoublets, addIterativeLSI, addClusters.. and others  in the downstream are getting the output to the same object name 'proj.'  When I ran the code first time, I input 'proj' that I got in the just above function. After reading the code lines again, I think I was wrong.  The input 'proj' is from 'ArchRProject or 'filterDoublets.'  Could you tell me if I understood correctly?

Thank you, 
Sanghoon

**Comment by rcorces** [comment, marked as answer]
This is very typical of the `R` programming language and is a bit counterintuitive for people coming from different languages.

The return value of most functions in ArchR is an ArchRProject object.

So you pass your current project to the function, it effectively creates a copy of that object while it performs operations, and then it returns the updated object which we use to overwrite the original object.

**Reply by sanghoonleeeins** [reply]
Thank you for your answer Ryan! 'proj' is overwritten and updated as I ran the downstream code.
